kame/kame/mping/mpsend.8
.\"	$KAME: mpsend.8,v 1.2 2000/11/09 03:04:09 itojun Exp $
.\"
.\" Copyright (C) 1999 WIDE Project.
.\" All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the project nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd Nov 13, 1999
.Dt MPSEND 1
.Os KAME
.\"
.Sh NAME
.Nm mpsend
.Nd a daemon program to send multicast udp packets
.\"
.Sh SYNOPSIS
.Nm mpsend
.Op Fl c Ar count
.Op Fl l Ar hop_limit
.Op Fl p Ar port
.Op Fl s Ar size
.Op Fl t Ar interval
.Op Fl v
.Fl i Ar interface
.Fl m Ar mcastaddr
.\"
.Sh DESCRIPTION
The
.Nm
program is designed for measure multicast routing stability in IPv6
network used with
.Xr mprecv 8
program.
The
.Nm
sends UDP packets to the multicast destination
.Ar mcastaddr
using
.Ar interface
in daemon mode.
.Pp
The options are as follows:
.Bl -tag -width Fl
.It Fl c Ar count
Specify the number of packet to be sent before the 
.Nm
terminates.
By default, the
.Nm
sends packets infinitely.
.It Fl l Ar hop_limit
Specify the hop_limit value of the packets. Default value is 1.
.It Fl p Ar port
Specify the UDP port number. Default value is 9998.
.It Fl s Ar size
Specify the packet payload size in byte. IPv6/UDP headers are not
included. Default value is 128 and minimum value is 16.
.It Fl t Ar interval
Specify the interval of packets in second. Default value is 10.
Note this value can be a floating point value, such as 0.1sec.
Only the super user can specify smaller value than 1sec.
.It Fl v
Enable the debug output and prevent the
.Nm
go into daemon mode.
.El
.\"
.Sh RETURN VALUES
The
.Nm
exit if
.Ar count
is specified and
.Ar count
packets are sent.
.\"
.Sh SEE ALSO
.Xr mprecv 8
.\"
.Sh HISTORY
The
.Nm
command first appeared in WIDE/KAME Ipv6 protocol stack kit.
.\"
.Sh BUGS
IPv4 is not supported.
.Pp
The format of logfile may not be not readable.


CHANGELOG
$KAME: CHANGELOG,v 1.2458 2003/08/18 05:39:09 itojun Exp $

<200308>
Sun Aug 17 JST 2003  itojun@iijlab.net
	* sys/netinet{,6}/in{,6}_proto.c: enforce inbound IPsec policy on
	  protocols that falls into raw socket as a wildcard case.

2003-08-15  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/pim6sd/{cfparse.y, mrt.[ch], pim6_proto.c, rp.c}
	Now static-RP configuration can work together with bootstrap-
	message-based RP configuration

2003-08-10  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/{freebsd5, kame}
	sync with FreeBSD-5.1 RELEASE (behaviour is not confirmed yet)

2003-08-07  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* kame/kame/had/mpa.c,haadisc.c
	A home address should be passed in the source address of a mobile
	solicit pakcet. This problem was reported and donated the code by
	Kenichi Yajima <yajima@netlab.nec.co.jp>
	
2003-08-07  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/cftoken.l: use a stronger lexical check for
	interface names.
	From: Hajimu UMEMOTO <ume@mahoroba.org>
	* kame/kame/dhcp6/config.c (configure_interface): check interfaces
	specified in the configuration file really exist.

2003-08-05  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* removed vrrp6 support because it might be a patent infringement.
	  http://www.ietf.org/ietf/IPR/cisco-ipr-draft-ietf-vrrp-ipv6-spec.txt

2003-08-01  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: implemented various clarifications and some new
	  features:
	- supported DNS name and NTP servers options
	- dhcp6s now uses the server/relay port to send a relay-rely
	  message
	- dhcp6s now ignores a rebind message when it cannot find a
	  binding, according to prefix-delegation-04
	- dhcp6relay now accepts relay-reply messages on the socket
	  listening on the server/relay port
	* The DHCPv6 implementation is now built by default under
	  *bsd/usr.sbin/.

<200307>
Thu Jul 31 19:19:48 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c,mip6_cncore.c
	sending/receiving a binding refresh request message is supported.

Tue Jul 29 17:09:15 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.[hc]
	select a CoA using the source address selection like algorithm.
	suggested by Francis.Dupont@enst-bretagne.fr.

2003-07-28  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* sys/netinet6/{ip6_output.c, ip6_var.h, mip6_cncore.c,
	mip6_hacore.[ch], nd6.c}:
	CN or HA dosen't add a type2 routing header to a Binding
	Acknowledgement packet when a deregistration packet was issued
	on home link and it was failed by some reasons.
	reported by Yukiyo.Akisada@jp.yokogawa.com and v6pc cert.
	team.

Mon Jul 28 20:03:46 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_cncore.c,mip6_mncore.[hc],in6.c,nd6_rtr.c
	  kame/sys/net/if_hif.[hc]
	assign a CoA separately for each home interface, since a home
	interface is logically independent from each other.

Mon Jul 28 14:38:11 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_hacore.c,mip6_cncore.c:
	- fixed a bug not sending a binding ack when DAD for a link-local
	  address is failed.  (en-bugged during re-structuring BC logic.)
	- fixed a bug using RTHDR2 with Mobility Headers other than
	  a binding ack.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

Fri Jul 25 19:12:15 JST 2003  itojun@iijlab.net
	* sys/netinet/ah_core.c: hmac-ripemd160 support

Fri Jul 25 18:07:43 JST 2003 sakane@kame.net
	* kame/sys/netkey/key.c
	fixed that the kernel crashed when key_spdacquire() was called
	because key_spdacquire() had been implemented imcopletely.

Thu Jul 24 16:11:06 JST 2003	keiichi@iij.ad.jp
	* mip6 related files.
	simplify the prefix and advertising router list management
	mechanism on a mobile node.

2003-07-20  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/config.c (configure_duid): fixed a bug of
	allocating short memory.  Upgrading is required if you configure
	dhcp6s with the duid statement.

Sat Jul 19 18:12:01 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_aesctr.c: support draft-ietf-ipsec-ciph-aes-ctr-03
	  as there's no official DOI assignment yet, we use private DOI number.
	* sys/netinet6/ah_aesxcbcmac.c: support
	  draft-ietf-ipsec-ciph-aes-xcbc-mac-03.
	  as there's no official DOI assignment yet, we use private DOI number.

Tue Jul 15 20:02:05 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_rijndael.c: simplify and update rijndael code.
	  markus@openbsd

2003-07-14  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: revised the relay agent implementation
	based on dhcpv6-28.
	- dhcp6relay now relays packets between clients and servers using
	  relay forward/reply messages.
	- dhcp6s now accepts relay forward messages and replies with relay
	  reply messages.

Sat Jul 12 15:05:23 JST 2003	suz@crl.hitachi.co.jp
	* openbsd/sys/netinet/igmp_var.h
	makes IGMPV3-kernel compilable on openbsd.

Thu Jul 10 21:35:04 JST 2003	keiichi@iij.ad.jp
	* kame/kame/had/halist.c
	fixed a bug that a DHAAD reply message only includes one global
	address per home agent.  the message must include all global
	addresses of all home agents.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

2003/07/10 21:07:50 JST kjc@csl.sony.co.jp
	the first step for transition to pf/altq.
	marge pf/altq into the existing altq-3.
	allow both pf/altq and altq-3 to coexit.
	currently, ALTQ3_COMPAT and ALTQ3_CLFIER_COMPAT are defined
	in sys/altq/altq.h to support altq-3.
	altq-3 will remain in kame for research experiments but will
	not be merged into bsd releases.

2003-07-08  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* kame/sys/netinet6/{in6_ifattach.c,ip6_mroute.c,ip6_mroute.h}
	protect against interface removal.

Tue Jul  8 18:51:06 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c
	insert alternate careof address sub-option when creating a
	binding update for the home registration.  This sub-option is
	MUST when we protect the message with ESP and strictly speaking,

kame/kame/mping/mpsend.c
/*	$KAME: mpsend.c,v 1.4 2000/12/05 09:08:54 itojun Exp $	*/

/*
 * Copyright (C) 1999 WIDE Project.
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the project nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include "mping.h"

double interval = DEFAULT_INTERVAL;
int size = DEFAULT_SIZE;
int hlim = DEFAULT_HOPLIMIT;
char *port = DEFAULT_PORT;
char *ifname = NULL;		/* no default interface this time */
char *maddr = NULL;		/* no default multicast address this time */
int count = 0;
int verbose = 0;
char *argv0;

void usage __P((void));

int
main(argc, argv)
	int argc;
	char **argv;
{
	int ch;
	int s, error;
	int int_sec, int_usec;
	u_char *buf, *p;
	struct addrinfo hints, *res;
	struct timeval tp;
	struct mping *mp;

	argv0 = *argv;
	while ((ch = getopt(argc, argv, "c:i:l:m:p:t:s:v")) != -1)
		switch (ch) {
		case 'c':
			count = atoi(optarg);
			break;
		case 'i':
			ifname = optarg;
			break;
		case 'l':
			hlim = atoi(optarg);
			break;
		case 'm':
			maddr = optarg;
			break;
		case 'p':
			port = optarg;
			break;
		case 't':
			interval = atof(optarg);
			break;
		case 's':
			size = atoi(optarg);
			break;
		case 'v':
			verbose++;
			break;
		default:
			usage();
			exit(-1);
		}
	argc -= optind;
	argv += optind;

	if (maddr == NULL) {
		fprintf(stderr, "need to specify destination address\n");
		exit(-1);
	}
	if (size < sizeof(struct mping))
		size = sizeof(struct mping);
	if (size > MAX_MSGSIZE)
		size = MAX_MSGSIZE;

	if (verbose) {
		fprintf(stderr, "%s current config is as follows:\n", argv0);
		fprintf(stderr, "\tifname:      %s\n", ifname);
		fprintf(stderr, "\tmaddr:       %s\n", maddr);
		fprintf(stderr, "\thlim:        %d\n", hlim);
		fprintf(stderr, "\tport:        %s\n", port);
		fprintf(stderr, "\tinterval:    %f\n", interval);
		fprintf(stderr, "\tsize:        %d\n", size);
		fprintf(stderr, "\tcount:       %d\n", count);
	}

	if (interval < 1.0 && getuid() != 0) {
		fprintf(stderr, "Only superuser can specify interval smaller than 1 second\n");
		exit(-1);
	}

	memset(&hints, 0, sizeof(hints));
	hints.ai_family = PF_INET6;
	hints.ai_socktype = SOCK_DGRAM;
	if ((error = getaddrinfo(maddr, port, &hints, &res)) != 0)
		err(1, "%s", gai_strerror(error));
	if (verbose) {
		u_char *p;
		int i;

		fprintf(stderr, "\tai_family:   %d\n", res->ai_family);
		fprintf(stderr, "\tai_socktype: %d\n", res->ai_socktype);
		fprintf(stderr, "\tai_protocol: %d\n", res->ai_protocol);
		fprintf(stderr, "\tai_addrlen:  %d\n", res->ai_addrlen);
		fprintf(stderr, "\tai_addr:     ");
		for (p = (u_char *)res->ai_addr, i = res->ai_addrlen; i;
			i--, p++) {
			fprintf(stderr, "%02x ", *p);
			if (i == 21)
				fprintf(stderr, "\n\t\t     ");
		}
		fprintf(stderr, "\n");
	}

	if ((s = socket(res->ai_family, res->ai_socktype, res->ai_protocol))
		< 0) 
		err(1, "%s", strerror(errno));

	if (IN6_IS_ADDR_MULTICAST(&(((struct sockaddr_in6 *)(res->ai_addr))->sin6_addr))) {
		int ifindex;

		if (ifname == NULL) {
			fprintf(stderr, "need to specify ifname for multicast\n");
			exit(-1);
		}
		ifindex = if_nametoindex(ifname);
		if (verbose)
			fprintf(stderr, "\tifindex   : %d\n", ifindex);
		error = setsockopt(s, IPPROTO_IPV6, IPV6_MULTICAST_IF,
				&ifindex, sizeof(ifindex));
		if (error < 0)
			err(1, "setsockopt(IPV6_MULTICAST_IF)");
		error = setsockopt(s, IPPROTO_IPV6, IPV6_MULTICAST_HOPS,
			&hlim, sizeof(hlim));
		if (error < 0)
			err(1, "setsockopt(IPV6_MULTICAST_HOPS)");
	}
	if ((buf = (u_char *)malloc(size)) == NULL)
		err(1, NULL);
	mp = (struct mping *)buf;

	if (gettimeofday(&tp, NULL) < 0)
		err(1, NULL);
	srandom(tp.tv_sec + tp.tv_usec);
	mp->m_sessid = random();
	fprintf(stderr, "session id: %ld (%08lx)\n",
		mp->m_sessid, mp->m_sessid);
	mp->m_seq = 0;
	for (p = buf + sizeof(struct mping); p < buf + size; p++)
		*p = (buf - p) & 0xff;
	int_sec = (int)interval;
	int_usec = (int)((interval - (double)int_sec) * 1000000);
	mp->m_interval.tv_sec = int_sec;
	mp->m_interval.tv_usec = int_usec;

	if (verbose == 0)
		daemon(0, 0);

	while (1) {
		if (verbose)
			fprintf(stderr, "sending %d bytes\n", size);
		error = sendto(s, buf, size, 0, res->ai_addr, res->ai_addrlen);
		if (error < 0)
			err(1, "sendto");
		if (count > 0 && --count == 0)
			break;
		mp->m_seq++;
		if (int_sec > 0)
			sleep(int_sec);
		if (int_usec > 0)
			usleep(int_usec);
	}
	freeaddrinfo(res);
	exit(0);
	/*NOTREACHED*/
}

void
usage()
{
	fprintf(stderr,
"Usage: %s [-v] [-c count] [-i interface] [-l hoplimit] [-p port] \n"
"        [-s size] [-t interval] -m mcastaddr\n", argv0);
}
